File: src/test/java/nz/co/searchwellington/utils/UrlCleanerTests.java
package nz.co.searchwellington.utils;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import nz.co.searchwellington.urls.shorturls.CachingUrlResolverService;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

public class UrlCleanerTests {
	
	private static final String SECURE_URL = "https://www.secure/rss";
	private static final String URL_FROM_USERLAND = " www.capitalshakers.co.nz/news/genesis-energy-shakers-vs-otago-rebels-easter-sunday/?PHPSESSID=eb1ca7999e13201d6dbf02a02fffacdd  ";
	private static final String BASIC_CLEANED_URL = "http://www.capitalshakers.co.nz/news/genesis-energy-shakers-vs-otago-rebels-easter-sunday/?PHPSESSID=eb1ca7999e13201d6dbf02a02fffacdd";
	private static final String EXPECTED_CLEAN_URL = "http://www.capitalshakers.co.nz/news/genesis-energy-shakers-vs-otago-rebels-easter-sunday/";

	private static final String SHORT_URL = "http://short.url/12234";
	private static final String UNCLEANED_SHORT_URL = " http://short.url/12234  ";
	
	@Mock CachingUrlResolverService urlResolverService;
	
	UrlCleaner cleaner;
		
	@Before
	public void setUp() throws Exception {
		MockitoAnnotations.initMocks(this);		
		when(urlResolverService.resolveUrl(BASIC_CLEANED_URL)).thenReturn(BASIC_CLEANED_URL);
		when(urlResolverService.resolveUrl(SHORT_URL)).thenReturn(BASIC_CLEANED_URL);
		when(urlResolverService.resolveUrl(SECURE_URL)).thenReturn(SECURE_URL);
		cleaner = new UrlCleaner(urlResolverService); 
	}
    
	@Test
    public void shouldStripPhpSessionId() throws Exception {
       assertEquals(EXPECTED_CLEAN_URL, cleaner.cleanSubmittedItemUrl(URL_FROM_USERLAND));
    }
    
	@Test
    public void shouldConsultRedirectingUrlResolvers() throws Exception {
    	String cleanedUrl = cleaner.cleanSubmittedItemUrl(SHORT_URL);
    	verify(urlResolverService).resolveUrl(SHORT_URL);
    	assertEquals(EXPECTED_CLEAN_URL, cleanedUrl);
    }
	
	@Test
    public void shouldCleanupShortUrlsBeforeResolving() throws Exception {
    	String cleanedUrl = cleaner.cleanSubmittedItemUrl(UNCLEANED_SHORT_URL);
    	verify(urlResolverService).resolveUrl(SHORT_URL);
    	assertEquals(EXPECTED_CLEAN_URL, cleanedUrl);
    }
	
	@Test
	public void testShouldPreappenedHttpPrefixIfMissing() throws Exception {
		String cleanedUrl = cleaner.cleanSubmittedItemUrl(UNCLEANED_SHORT_URL);
		assertEquals(EXPECTED_CLEAN_URL, cleanedUrl);
	}
	
	@Test
	public void testShouldAllowHttpsPrefix() throws Exception {
		String cleanedUrl = cleaner.cleanSubmittedItemUrl(SECURE_URL);
		assertEquals(SECURE_URL, cleanedUrl);

	}
	
}

File: src/test/scala/nz/co/searchwellington/urls/UrlCleanerTests.scala
package nz.co.searchwellington.urls

import nz.co.searchwellington.urls.shorturls.CachingShortUrlResolverService
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeEach
import org.mockito.Mockito.{mock, verify, when}

import java.net.URI

class UrlCleanerTests {

  private val SECURE_URL = new URI("https://www.secure/rss")
  private val URL_FROM_USERLAND = " www.capitalshakers.co.nz/news/genesis-energy-shakers-vs-otago-rebels-easter-sunday/?PHPSESSID=eb1ca7999e13201d6dbf02a02fffacdd  "
  private val BASIC_CLEANED_URL = new URI("http://www.capitalshakers.co.nz/news/genesis-energy-shakers-vs-otago-rebels-easter-sunday/?PHPSESSID=eb1ca7999e13201d6dbf02a02fffacdd")
  private val EXPECTED_CLEAN_URL = "http://www.capitalshakers.co.nz/news/genesis-energy-shakers-vs-otago-rebels-easter-sunday/"
  private val SHORT_URL = new URI("http://short.url/12234")
  private val UNCLEANED_SHORT_URL = " http://short.url/12234  "
  
  private val shortUrlResolverService = mock(classOf[CachingShortUrlResolverService])
  private val cleaner = new UrlCleaner(shortUrlResolverService)

  @BeforeEach
  def setUp(): Unit = {
    when(shortUrlResolverService.resolveUrl(BASIC_CLEANED_URL)).thenReturn(BASIC_CLEANED_URL)
    when(shortUrlResolverService.resolveUrl(SHORT_URL)).thenReturn(BASIC_CLEANED_URL)
    when(shortUrlResolverService.resolveUrl(SECURE_URL)).thenReturn(SECURE_URL)
  }

  @Test
  def shouldStripPhpSessionId(): Unit = assertEquals(EXPECTED_CLEAN_URL, cleaner.cleanSubmittedItemUrl(URL_FROM_USERLAND))

  @Test
  def shouldConsultRedirectingUrlResolvers(): Unit = {
    val cleanedUrl = cleaner.cleanSubmittedItemUrl(SHORT_URL.toURL.toExternalForm)
    verify(shortUrlResolverService).resolveUrl(SHORT_URL)
    assertEquals(EXPECTED_CLEAN_URL, cleanedUrl)
  }

  @Test
  def shouldCleanupShortUrlsBeforeResolving(): Unit = {
    val cleanedUrl = cleaner.cleanSubmittedItemUrl(UNCLEANED_SHORT_URL)
    verify(shortUrlResolverService).resolveUrl(SHORT_URL)
    assertEquals(EXPECTED_CLEAN_URL, cleanedUrl)
  }

  @Test
  def testShouldPrependHttpPrefixIfMissing(): Unit = {
    val cleanedUrl = cleaner.cleanSubmittedItemUrl(UNCLEANED_SHORT_URL)
    assertEquals(EXPECTED_CLEAN_URL, cleanedUrl)
  }

  @Test
  def testShouldAllowHttpsPrefix(): Unit = {
    val cleanedUrl = cleaner.cleanSubmittedItemUrl(SECURE_URL.toURL.toExternalForm)
    assertEquals(SECURE_URL.toURL.toExternalForm, cleanedUrl)
  }

}